Non-blocking row_sparse_pull. Fix incorrect indices generated by device kvstore.row_sparse_pull

eric-haibin-lin opened the pull request:
## Description ##
This PR adds async execution support for kv.row_sparse_pull. 
The operation was blocking because it requires unique row_ids, whose shape cannot be inferred ahead of time. This PR stores the unique row_ids in a row_sparse NDArray, whose data shape can be changed at run time when executed asynchronously. 


## Checklist ##
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Removed `use_copy` param in `BroadcastRowSparse ` - this is essentially the same as calling `Broadcast`
- Removed `CopyRetainedRowsToGPU ` and always use `SparseRetain` because `CopyRetainedRowsToGPU` has high invocation overhead and `SparseRetain` has improved performance
- Revised test cases to test against shapes/dtypes commonly used by users
- Tested example/sparse/linear_classification with dist_sync kvstore
- Tested tests/nightly/dist_sync_kvstore.py
- Tested tests/python/gpu/test_kvstore.py
- Tested tests/python/unittest/test_kvstore.py

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

eric-haibin-lin commented on src/kvstore/kvstore_dist.h:
```diff
@@ -279,24 +279,20 @@ class KVStoreDist : public KVStoreLocal {
+        target_val_rowids[i].second = Unique(row_id, pinned_ctx_, 0);
       }
+      CHECK_EQ(num_vals, 1) << "RowSparsePull with multiple values is not supported yet";
+      NDArray& indices = target_val_rowids[0].second;
```
It should be fine, they're passed by value when pushed to engine

eric-haibin-lin commented:
@reminisce could you help review?

rahul003 commented on src/kvstore/kvstore_utils.cc:
```diff
@@ -28,15 +28,18 @@
+    IType *dptr = out.data().dptr<IType>();
+    common::ParallelSort(dptr, dptr + num_elements,
+                         engine::OpenMP::Get()->GetRecommendedOMPThreadCount());
+    const size_t num_selected_out = std::unique(dptr, dptr + num_elements) - dptr;
```
`std::unique` only removes duplicates of elements present consecutively. Is that what's expected?

eric-haibin-lin replied:
Yes because inputs are sorted first

reminisce commented on tests/python/unittest/test_kvstore.py:
```diff
@@ -76,7 +76,7 @@ def check_row_sparse_pull(kv, count):
             vals.append(mx.nd.zeros(shape).tostype('row_sparse'))
             row_id = np.random.randint(num_rows, size=num_rows)
-            row_ids.append(mx.nd.array(row_id))
+            row_ids.append(mx.nd.array(row_id).reshape((2,2)))
```
Why reshape?
One suggestion: (2, 2) is too hard-coded. If shape is changed in the beginning, the test would fail.

eric-haibin-lin replied:
You are right.. Made it less hard-coded by calculating the dimension based on total number of elements. 
I'm testing 2-D rowids explicitly because that seems to be a common use case, many users forget to reshape it to 1-D before passing it to kvstore. It is now supported with this PR.

eric-haibin-lin commented:
This PR also fixed a bug in the original GPU "Unique" implementation, where the same pointer is passed as both the input/output to cub::unique. This results in undefined output when the number of rowids is large. Fixed by storing the input/output in separate buffers.

eric-haibin-lin commented:
@ZiyueHuang please view the changes for gpu unique

ZiyueHuang commented:
@eric-haibin-lin Thanks for the fix!

Looks good.
